File: shared/src/commonMain/kotlin/com/snabbit/runner/shared/features/tiering/presentation/TierNudgeWeekFlags.kt
package com.snabbit.runner.shared.features.tiering.presentation

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Outline
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.snabbit.design.atoms.SnabbitText
import com.snabbit.runner.shared.core.designsystem.TierColors
import com.snabbit.runner.shared.features.tiering.domain.model.Tier
import com.snabbit.runner.shared.features.tiering.domain.model.TierCoinsData
import com.snabbit.runner.shared.features.tiering.domain.model.TierWeek

/**
 * A centred row of up to [maxFlags] (≤4) "Week N" flags — one per week in the
 * `tier_nudge` coins body ([data]). Each flag is the pennant ([WeekFlagShape],
 * ported from the Flutter `week_coin_flag.svg`) filled with the runner's [tier]
 * colour, with "Week N" centred in white. The current week (`current_week`) is
 * full opacity; preceding weeks are dimmed to 30%. When `current_week` is unknown
 * the whole row stays full opacity (mirrors the Flutter fix — no all-dimmed row).
 *
 * When there are more than [maxFlags] weeks the row is windowed so the current
 * week stays visible (matching the Flutter `_windowed`).
 */
@Composable
fun TierNudgeWeekFlags(
    data: TierCoinsData,
    tier: Tier,
    modifier: Modifier = Modifier,
    maxFlags: Int = MAX_FLAGS,
    strings: TieringStrings = rememberTieringStrings(),
) {
    val coins = data.coins
    val weeks = coins?.weeks.orEmpty()
    if (weeks.isEmpty()) return
    val currentWeek = coins?.currentWeek
    val visible = windowed(weeks, currentWeek, maxFlags.coerceIn(0, MAX_FLAGS))

    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(4.dp, Alignment.CenterHorizontally),
        verticalAlignment = Alignment.Top,
    ) {
        visible.forEach { week ->
            WeekFlag(
                weekNumber = week.week,
                tier = tier,
                // Highlight the current week; when current_week is unknown keep the
                // whole row at full opacity instead of dimming every flag to 30%.
                current = currentWeek == null || week.week == currentWeek,
                weekWord = strings.weekWord,
            )
        }
    }
}

/**
 * At most [maxFlags] flags. With more weeks than that, window so the current week
 * stays visible (the [maxFlags] weeks ending at it); if the current week isn't
 * found, fall back to the first [maxFlags] in order.
 */
private fun windowed(weeks: List<TierWeek>, currentWeek: Int?, maxFlags: Int): List<TierWeek> {
    if (weeks.size <= maxFlags) return weeks
    val idx = if (currentWeek == null) -1 else weeks.indexOfFirst { it.week == currentWeek }
    if (idx < 0) return weeks.take(maxFlags)
    val end = if (idx + 1 < maxFlags) maxFlags else idx + 1
    return weeks.subList(end - maxFlags, end)
}

@Composable
private fun WeekFlag(weekNumber: Int?, tier: Tier, current: Boolean, weekWord: String) {
    Box(
        modifier = Modifier
            .width(68.dp)
            .height(20.dp)
            .alpha(if (current) 1f else 0.3f)
            .background(TierColors.weekFlag(tier), WeekFlagShape),
        contentAlignment = Alignment.Center,
    ) {
        SnabbitText(
            text = "$weekWord ${weekNumber ?: ""}".trim(),
            fontSize = 10.sp,
            fontWeight = FontWeight.Medium,
            color = TierColors.onAccent,
        )
    }
}

private const val MAX_FLAGS = 4

/** The pennant outline (rounded top, scalloped bottom), scaled from the 68×19 Flutter SVG. */
private val WeekFlagShape = object : Shape {
    override fun createOutline(
        size: Size,
        layoutDirection: LayoutDirection,
        density: Density,
    ): Outline {
        val sx = size.width / 68f
        val sy = size.height / 19f
        fun px(v: Float) = v * sx
        fun py(v: Float) = v * sy
        val path = Path().apply {
            moveTo(px(0f), py(0f))
            lineTo(px(68f), py(0f))
            cubicTo(px(65.7909f), py(0f), px(64f), py(1.79086f), px(64f), py(4f))
            lineTo(px(64f), py(16.7308f))
            cubicTo(px(64f), py(18.2116f), px(62.422f), py(19.1584f), px(61.1154f), py(18.4615f))
            lineTo(px(60.25f), py(18f))
            cubicTo(px(57.9062f), py(16.75f), px(55.0938f), py(16.75f), px(52.75f), py(18f))
            cubicTo(px(50.4062f), py(19.25f), px(47.5938f), py(19.25f), px(45.25f), py(18f))
            cubicTo(px(42.9062f), py(16.75f), px(40.0938f), py(16.75f), px(37.75f), py(18f))
            cubicTo(px(35.4062f), py(19.25f), px(32.5938f), py(19.25f), px(30.25f), py(18f))
            cubicTo(px(27.9062f), py(16.75f), px(25.0938f), py(16.75f), px(22.75f), py(18f))
            cubicTo(px(20.4062f), py(19.25f), px(17.5938f), py(19.25f), px(15.25f), py(18f))
            cubicTo(px(12.9062f), py(16.75f), px(10.0938f), py(16.75f), px(7.75f), py(18f))
            lineTo(px(6.88462f), py(18.4615f))
            cubicTo(px(5.57802f), py(19.1584f), px(4f), py(18.2116f), px(4f), py(16.7308f))
            lineTo(px(4f), py(4f))
            cubicTo(px(4f), py(1.79086f), px(2.20914f), py(0f), px(0f), py(0f))
            close()
        }
        return Outline.Generic(path)
    }
}

File: lib/widgets/tiering/tier_nudge_week_flags.dart
import 'package:flutter/material.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';
import 'package:flutter_svg/flutter_svg.dart';
import 'package:snabbit_runner/constants/assets_constants.dart';
import 'package:snabbit_runner/models/tiering/tier_coins_data.dart';
import 'package:snabbit_runner/utils/colors.dart';
import 'package:snabbit_runner/utils/enums.dart';

/// A centred row of up to four "Week N" flags — one per week in the coins body
/// ([coins]) supplied by the parent [TierNudgeProgressCard]. Each flag is the
/// [AssetConstants.weekCoinFlag] pennant tinted with the runner's [tier] colour,
/// with "Week N" centred in white. The current week (`current_week`) is full
/// opacity; preceding weeks are dimmed to 30%. When the current week is unknown
/// the whole row stays at full opacity rather than looking disabled.
///
/// Data is passed down from the card (which already parsed it) — this widget
/// does not re-read providers or re-parse the payload.
class TierNudgeWeekFlags extends StatelessWidget {
  const TierNudgeWeekFlags({
    super.key,
    required this.coins,
    required this.tier,
  });

  /// Weekly coins progress (weeks + `current_week`), from the parent card's data.
  final TierCoinsProgress? coins;

  /// Runner tier — drives the pennant colour.
  final Tier? tier;

  static const int _maxFlags = 4;
  static const double _flagWidth = 68;
  static const double _flagHeight = 20;

  @override
  Widget build(BuildContext context) {
    final weeks = coins?.weeks ?? const <TierWeek>[];
    if (weeks.isEmpty) return const SizedBox.shrink();

    final currentWeek = coins?.currentWeek;
    final color = _flagColor(tier);
    final visible = _windowed(weeks, currentWeek);

    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        for (final week in visible)
          _WeekFlag(
            week: week.week,
            color: color,
            // Highlight the current week. When `current_week` is unknown, keep
            // the whole row at full opacity instead of dimming every flag to 30%.
            isCurrent: currentWeek == null || week.week == currentWeek,
          ),
      ],
    );
  }

  /// At most [_maxFlags] flags. When there are more weeks than that, window so
  /// the current week stays visible (the [_maxFlags] weeks ending at it); if the
  /// current week isn't found, fall back to the first [_maxFlags] in order.
  List<TierWeek> _windowed(List<TierWeek> weeks, int? currentWeek) {
    if (weeks.length <= _maxFlags) return weeks;
    final idx = currentWeek == null
        ? -1
        : weeks.indexWhere((w) => w.week == currentWeek);
    if (idx < 0) return weeks.take(_maxFlags).toList();
    final end = idx + 1 < _maxFlags ? _maxFlags : idx + 1;
    return weeks.sublist(end - _maxFlags, end);
  }
}

/// Per-tier flag background colour; legacy/unknown tiers fall back to base.
Color _flagColor(Tier? tier) {
  switch (tier) {
    case Tier.BASE:
      return AppColors.tierWeekFlagBase;
    case Tier.SILVER:
      return AppColors.tierWeekFlagSilver;
    case Tier.GOLD:
      return AppColors.tierWeekFlagGold;
    case Tier.DIAMOND:
      return AppColors.tierWeekFlagDiamond;
    case Tier.PINK_DIAMOND:
      return AppColors.tierWeekFlagPinkDiamond;
    case Tier.BASIC:
    case Tier.PRO:
    case Tier.ELITE:
    case null:
      return AppColors.tierWeekFlagBase;
  }
}

/// A single pennant: the tinted flag SVG with "Week N" centred, dimmed to 30%
/// unless it is the current week.
class _WeekFlag extends StatelessWidget {
  const _WeekFlag({
    required this.week,
    required this.color,
    required this.isCurrent,
  });

  final int? week;
  final Color color;
  final bool isCurrent;

  @override
  Widget build(BuildContext context) {
    return Opacity(
      opacity: isCurrent ? 1.0 : 0.3,
      child: SizedBox(
        width: TierNudgeWeekFlags._flagWidth.w,
        height: TierNudgeWeekFlags._flagHeight.h,
        child: Stack(
          alignment: Alignment.center,
          children: [
            Positioned.fill(
              child: SvgPicture.asset(
                AssetConstants.weekCoinFlag,
                fit: BoxFit.fill,
                colorFilter: ColorFilter.mode(color, BlendMode.srcIn),
              ),
            ),
            Text(
              'Week ${week ?? ''}'.trim(),
              textAlign: TextAlign.center,
              style: TextStyle(
                color: AppColors.n0,
                fontSize: 10.sp,
                fontWeight: FontWeight.w500,
                letterSpacing: 0,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
